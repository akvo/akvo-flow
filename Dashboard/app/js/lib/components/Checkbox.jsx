import React from 'react';
import PropTypes from 'prop-types';

class Checkbox extends React.Component {
  render() {
    const { id, name, checked } = this.props;

    return (
      <input
        type="checkbox"
        id={id}
        name={name}
        checked={checked}
        onChange={() => this.props.onChange(name, !checked)}
        className="displayInline"
      />
    );
  }
}

Checkbox.defaultProps = {
  checked: false,
};

Checkbox.propTypes = {
  checked: PropTypes.bool,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Checkbox;
