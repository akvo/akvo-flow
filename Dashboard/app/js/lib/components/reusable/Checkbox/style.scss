/* Customize the label (the container) */
.checkbox-container {
  display: inline;
  position: relative;
  padding-left: 25px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;

    &:checked ~ .checkmark {
      background-color: #424242;

      &::after {
        display: block;
      }
    }
  }

  .checkmark {
    position: absolute;
    top: 0;
    left: 0;
    height: 16px;
    width: 16px;
    border-radius: 2px;
    border: 1px solid #424242;

    &:after {
      content: '';
      position: absolute;
      display: none;
      left: 5px;
      top: 0px;
      width: 5px;
      height: 9px;
      border: solid white;
      border-width: 0 2px 2px 0;
      -webkit-transform: rotate(45deg);
      -ms-transform: rotate(45deg);
      transform: rotate(45deg);
    }
  }
}
