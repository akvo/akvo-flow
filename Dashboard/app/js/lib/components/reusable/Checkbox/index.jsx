import React from 'react';
import PropTypes from 'prop-types';
import './style.scss';

class Checkbox extends React.Component {
  render() {
    const { id, name, checked, label } = this.props;

    // Label has no id because select all checkboxes has the same id of 0
    // so if two labels has the same id, when i click on one the other one gets checked too
    return (
      <label htmlFor="no_id" className="checkbox-container">
        {label}
        <input
          type="checkbox"
          id={id}
          name={name}
          checked={checked}
          onChange={() => this.props.onChange(name, !checked)}
          className="displayInline"
        />
        <span className="checkmark" />
      </label>
    );
  }
}

Checkbox.defaultProps = {
  checked: false,
};

Checkbox.propTypes = {
  checked: PropTypes.bool,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
};

export default Checkbox;
