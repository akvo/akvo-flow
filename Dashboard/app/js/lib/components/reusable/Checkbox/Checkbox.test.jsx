import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import Checkbox from './index';
import '@testing-library/jest-dom/extend-expect';

// https://github.com/testing-library/react-testing-library#suppressing-unnecessary-warnings-on-react-dom-168
// eslint-disable-next-line no-console
const originalError = console.error;
beforeAll(() => {
  // eslint-disable-next-line no-console
  console.error = (...args) => {
    if (/supports * the "act"/.test(args[0])) {
      return;
    }
    originalError.call(console, ...args);
  };
});

afterAll(() => {
  // eslint-disable-next-line no-console
  console.error = originalError;
});

// Main Tests
describe('Checkbox Tests', () => {
  afterEach(cleanup);

  const props = {
    id: 'fake-id',
    name: 'fake-id',
    checked: false,
    label: 'Fake checkbox',
    onChange: jest.fn(),
  };

  it('+++ renders <snapshot>', () => {
    const wrapper = render(<Checkbox {...props} />);

    expect(wrapper.container).toMatchSnapshot();
  });

  test('+++ checkbox works', () => {
    const wrapper = render(<Checkbox {...props} />);

    // find input and trigger a check
    const inputNode = wrapper.getByLabelText('Fake checkbox');
    fireEvent.click(inputNode);

    expect(props.onChange).toHaveBeenCalledTimes(1);
    expect(props.onChange).toHaveBeenCalledWith('fake-id', true);
  });
});
