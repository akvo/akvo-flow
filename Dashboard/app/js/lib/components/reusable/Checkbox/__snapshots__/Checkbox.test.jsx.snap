// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Tests +++ renders <snapshot> 1`] = `
<div>
  <label
    class="checkbox-container"
    for="809082"
  >
    Fake checkbox
    <input
      class="displayInline"
      id="809082"
      name="809082"
      type="checkbox"
    />
    <span
      class="checkmark"
    />
  </label>
</div>
`;
